lib/gc/Dialect/Linalgx/LinalgxOps.cpp

```diff
@@ -82,6 +82,385 @@ void SigmoidOp::getEffects(
 }
 
+//===----------------------------------------------------------------------===//
+// Mmt2DVnniOp
+//===----------------------------------------------------------------------===//
+
+SmallVector<utils::IteratorType> Mmt2DVnniOp::getIteratorTypesArray() {
+  return SmallVector<utils::IteratorType>{
+      utils::IteratorType::parallel,  utils::IteratorType::parallel,
+      utils::IteratorType::reduction, utils::IteratorType::reduction,
+      utils::IteratorType::reduction, utils::IteratorType::reduction};
+}
+
+static SmallVector<AffineExpr> getSymbolBindings(Mmt2DVnniOp self) {
+  MLIRContext *context = self.getContext();
+
+  auto vnniShape = ShapeAdaptor(self.getInputs()[1].getType());
+
+  SmallVector<AffineExpr> exprs;
+  exprs.push_back(getAffineSymbolExpr(0, context));
+  exprs.push_back(getAffineSymbolExpr(1, context));
+
+  int64_t cst2 = vnniShape.getDimSize(3);
+  exprs.push_back(getAffineConstantExpr(cst2, context));
+
+  exprs.push_back(getAffineSymbolExpr(3, context));
+
+  int64_t cst4 = vnniShape.getDimSize(2);
+  exprs.push_back(getAffineConstantExpr(cst4, context));
+
+  int64_t cst5 = vnniShape.getDimSize(4);
+  exprs.push_back(getAffineConstantExpr(cst5, context));
+  return exprs;
+}
+
+ArrayAttr Mmt2DVnniOp::getIndexingMaps() {
+  static const char memoizeAttr[] = "linalg.memoized_indexing_maps";
+  ArrayAttr cached = getOperation()->getAttrOfType<ArrayAttr>(memoizeAttr);
+  if (cached)
+    return cached;
+
+  static auto mapA = "affine_map<(d0, d1, d2, d3, d4, d5)[s0, s1, s2, s3, "
+                     "s4, s5] -> (d0, (d3 * s4 + d4) * s5 + d5)>";
+  static auto mapB = "affine_map<(d0, d1, d2, d3, d4, d5)[s0, s1, s2, s3, "
+                     "s4, s5] -> (d1, d3, d4, d2, d5)>";
+  static auto mapC = "affine_map<(d0, d1, d2, d3, d4, d5)[s0, s1, s2, s3, "
+                     "s4, s5] -> (d0, d1 * s2 + d2)>";
+  MLIRContext *context = getContext();
+  auto symbolBindings = getSymbolBindings(*this);
+  SmallVector<AffineMap> maps;
+  maps.push_back(llvm::cast<AffineMapAttr>(mlir::parseAttribute(mapA, context))
+                     .getValue());
+  maps.back() = simplifyAffineMap(
+      maps.back().replaceDimsAndSymbols({}, symbolBindings, 6, 0));
+  maps.push_back(llvm::cast<AffineMapAttr>(mlir::parseAttribute(mapB, context))
+                     .getValue());
+  maps.back() = simplifyAffineMap(
+      maps.back().replaceDimsAndSymbols({}, symbolBindings, 6, 0));
+  maps.push_back(llvm::cast<AffineMapAttr>(mlir::parseAttribute(mapC, context))
+                     .getValue());
+  maps.back() = simplifyAffineMap(
+      maps.back().replaceDimsAndSymbols({}, symbolBindings, 6, 0));
+  cached = Builder(context).getAffineMapArrayAttr(maps);
+  getOperation()->setAttr(memoizeAttr, cached);
+  return cached;
+}
+
+void Mmt2DVnniOp::regionBuilder(ImplicitLocOpBuilder &b, Block &block,
+                                ArrayRef<NamedAttribute> attrs) {
+  assert(3 > 0 && block.getNumArguments() == 3 &&
```

Comment:
redundant `3>0`?

Reply:
I reproduced the code structure from `mlir-linalg-ods-yaml-gen` codegen, just for consistency.